lasso/emme.py

```diff
@@ -268,145 +258,147 @@ def create_emme_network(
-
-    drive_links_df = links_df[
-        ~(links_df.A.isin(parameters.taz_N_list + parameters.maz_N_list)) & 
-        ~(links_df.B.isin(parameters.taz_N_list + parameters.maz_N_list)) &
-        ((links_df.drive_access == 1) & (links_df.assignable == 1))
-    ].copy()
-
-    model_tables["link_table"] = drive_links_df.to_dict('records')
-
-    drive_nodes_df = nodes_df[
-        (nodes_df.N.isin(drive_links_df.A.tolist()) + nodes_df.N.isin(drive_links_df.B.tolist()))
-    ].copy()
-
-    model_tables["node_table"] = drive_nodes_df.to_dict('records')
-
-    return model_tables
+    shapes = transit_network.feed["shapes"].copy()
+    trips = transit_network.feed["trips"].copy()
+    routes = transit_network.feed["routes"].copy()
+
+    bus_routes = routes.loc[routes["route_type"].isin([route_type_bus_id]), "route_id"]
+    bus_trips = trips.loc[trips["route_id"].isin(bus_routes), "shape_id"]
+    bus_shapes = shapes[shapes["shape_id"].isin(bus_trips)]
+    return bus_shapes
+
+def tag_if_link_contains_bus(links_df: gpd.GeoDataFrame, gtfs_shape_bus_routes: gpd.GeoDataFrame):
+    links_df = links_df.copy()
+    gtfs_shape_bus_routes = gtfs_shape_bus_routes.copy()
+
+    # links are defined as pairs of nodes, so get next node (this assumes order)
+    gtfs_shape_bus_routes["next_node_id"] = gtfs_shape_bus_routes["shape_model_node_id"].shift(1)
+    gtfs_shape_bus_routes = gtfs_shape_bus_routes.sort_values(by=["shape_id", "shape_pt_sequence"])
+    gtfs_shape_bus_routes = gtfs_shape_bus_routes[(gtfs_shape_bus_routes["shape_pt_sequence"] != 1)]
+    gtfs_shape_bus_routes = gtfs_shape_bus_routes.drop_duplicates(subset=["shape_model_node_id", "next_node_id"])
+    gtfs_shape_bus_routes["has_bus_on_link"] = True
+    gtfs_shape_bus_routes["shape_model_node_id"] = pd.to_numeric(gtfs_shape_bus_routes["shape_model_node_id"])
+    gtfs_shape_bus_routes["next_node_id"] = pd.to_numeric(gtfs_shape_bus_routes["next_node_id"])
+
+    links_df = pd.merge(links_df, gtfs_shape_bus_routes[["shape_model_node_id", "next_node_id", "has_bus_on_link"]], 
+        left_on=["A", "B"],
+        right_on=["shape_model_node_id", "next_node_id"],
+        how="left"
+    ).drop(columns=["shape_model_node_id", "next_node_id"])
+    links_df["has_bus_on_link"] = links_df["has_bus_on_link"].fillna(False)
+
+    return links_df
 
 def prepare_table_for_drive_network(
-    nodes_df,
-    links_df,
-    parameters,
+    nodes_df: pd.DataFrame,
+    links_df: pd.DataFrame,
+    transit_network: StandardTransit,
+    shapes_df: Optional[GeoDataFrame],
+    parameters: Parameters,
+    maximum_ft: int=7,
+    regenerate_connectors: bool=False,
 ):
 
     """
-    prepare model table for drive network, in which taz nodes are centroids
-    maz and tap nodes are included, but not as centroids
-    keep links that are drive_access == 1 and assignable == 1
+    prepare model table for drive network (including taz and maz nodes), in which taz/maz nodes are centroids
```

Comment:
MAZs should not be treated as centroids. As before, TAZs are the centroids, MAZs are in the network as regular links and nodes.

Reply:
my mistake I think I understand now, 
done d017d70